File: examples/java-threads/Example2.java
// =================================================================
//
// File: Example2.java
// Author: Pedro Perez
// Description: This file contains the code that searches for the
// 				smallest value stored in an array using using
//				Java's Threads.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================


public class Example2 extends Thread {
	private static final int SIZE = 100_000_000;
	private int array[], start, end;
	private int result;

	public Example2(int array[], int start, int end) {
		this.array = array;
		this.start = start;
		this.end = end;
		this.result = 0;
	}

	public int getResult() {
		return result;
	}

	public void run() {
		result = Integer.MAX_VALUE;
		for (int i = start; i < end; i++) {
			result = (int) Math.min(result, array[i]);
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int array[], block, min = 0;
		Example2 threads[];
		double ms;

		array = new int[SIZE];
		Utils.randomArray(array);
		Utils.displayArray("array", array);

		int pos = Math.abs(Utils.r.nextInt()) % SIZE;
		System.out.printf("Setting value 0 at %d\n", pos);
		array[pos] = 0;

		block = SIZE / Utils.MAXTHREADS;
		threads = new Example2[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int j = 1; j <= Utils.N; j++) {
			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = new Example2(array, (i * block), ((i + 1) * block));
				} else {
					threads[i] = new Example2(array, (i * block), SIZE);
				}
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);

			if (j == Utils.N) {
				min = Integer.MAX_VALUE;
				for (int i = 0; i < threads.length; i++) {
					min = (int) Math.min(min, threads[i].getResult());
				}
			}
		}
		System.out.printf("result = %d\n", min);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}

File: examples/07-tbb/example2.cpp
// =================================================================
//
// File: example2.cpp
// Author: Pedro Perez
// Description: This file contains the code that searches for the
// 				smallest value stored in an array using Intel's TBB.
//				To compile: g++ example2.cpp -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <climits>
#include <algorithm>
#include <tbb/parallel_reduce.h>
#include <tbb/blocked_range.h>
#include "utils.h"

const int SIZE = 1000000000; //1e9

using namespace std;
using namespace tbb;

class MinValue {
private:
	int *array, result;

public:
	MinValue(int *a) : array(a), result(INT_MAX) {}

	MinValue(MinValue &x, split) : array(x.array), result(INT_MAX) {}

	int getResult() const {
		return result;
	}

	void operator() (const blocked_range<int> &r) {
		for (int i = r.begin(); i != r.end(); i++) {
			result = min(result, array[i]);
		}
	}

	void join(const MinValue &x) {
		result = min(result, x.result);
	}
};

int main(int argc, char* argv[]) {
	int *a, pos, result;
	double ms;

	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("a", a);

	srand(time(0));
	pos = rand() % SIZE;
	printf("Setting value 0 at %i\n", pos);
	a[pos] = 0;

	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		start_timer();

		MinValue obj(a);
		parallel_reduce(blocked_range<int>(0, SIZE), obj);
		result = obj.getResult();

		ms += stop_timer();
	}
	cout << "result = " << result << endl;
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	delete [] a;
	return 0;
}
